### PR TITLE
Update _text_completion.py to support `packed` mode

### PR DESCRIPTION
**Pull request by andyl98**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
GitHub issue: https://github.com/pytorch/torchtune/issues/1058

#### Changelog
What are the changes made in this PR?
Add support for `packed` mode inside `_text_completion.py`

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add unit tests for any new functionality
- [ ] update docstrings for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

**Comment by RdoubleA** [comment]
Thanks for this update. Indeed, having packing for text completion datasets is almost required for continued pre-training. I purposely left this out because I was concerned by the memory cost of packing, which would be more problematic for text completion datasets that tend to be much larger. But for local text corpuses this should be okay. Are you trying to train on local data?

The code change looks good to me, do you mind just sharing a distributed run on text_completion_dataset both with and without packing? Curious to see if the performance gains are similar to instruct and chat, also good to ensure that you don't run into any OOMs :)

**Comment by andyl98** [comment]
> The code change looks good to me, do you mind just sharing a distributed run on text_completion_dataset both with and without packing? Curious to see if the performance gains are similar to instruct and chat, also good to ensure that you don't run into any OOMs :)

Sounds good, will update the results once a training run is done :) And yes, I'm training on local data with a small set (<1B tokens in total).

File: torchtune/datasets/_text_completion.py [hunk 1]
```diff
@@ -94,6 +97,7 @@ def text_completion_dataset(
         ...   column="text",
         ...   max_seq_len=2096,
         ...   data_dir="realnewslike",
+        ...   packed=True,
```
Comment (RdoubleA) [inline torchtune/datasets/_text_completion.py:100]: For this example, let's actually set packed=False because for this dataset most hardware setups will OOM if packed

File: torchtune/datasets/_text_completion.py [hunk 2]
```diff
@@ -104,14 +108,17 @@ def text_completion_dataset(
             column: text
             max_seq_len: 2096
             data_dir: realnewslike
+            packed: True
```
Comment (RdoubleA) [inline torchtune/datasets/_text_completion.py:111]: Same comment here

**Comment by RdoubleA** [comment]
Hi @andyl98, were you able to try launching a run with this change? If you're having trouble let me know, and I could also launch a run on my end and get this merged in.

**Comment by andyl98** [comment]
Screenshot 2024-06-10 at 7 47 18 PM
Hi @RdoubleA sorry wasn't able to work on the project. I tested with small dataset and the results look as expected. However I do recognize the memory consumption issue with a larger dataset. Hopefully the mmap dataset feature will come soon!

**Comment by RdoubleA** [comment]
Appreciate you launching a test run! Looks good to me and it's great to see a similar bump in QPS for text completion.

Yes, I am working on a more memory efficient dataset implementation, hopefully that will unlock using packing with any size dataset 😎

I'll run the CI and if there's no other issues get this merged. Thanks again!

**Review by RdoubleA — APPROVED** [review]
